clients/src/test/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumerTest.java

```diff
@@ -204,6 +206,42 @@ public void testAssignOnEmptyTopicInPartition() {
         assertThrows(IllegalArgumentException.class, () -> consumer.assign(singleton(new TopicPartition("  ", 0))));
     }
 
+    @Test
+    public void testBeginningOffsetsFailsIfNullPartitions() {
+        consumer = newConsumer(time, new StringDeserializer(), new StringDeserializer());
+        assertThrows(NullPointerException.class, () -> consumer.beginningOffsets(null,
```

Comment:
Can we not throw a NPE? Maybe `IllegalArgumentException`

Reply:
We could definitely do that, but just curious, why not NPE? I went with NPE for 2 reasons: first and foremost, because I wanted to use `Objects.requireNonNull` to validate the argument (which I find very neat), and that throws NPE instead of IllegalArg. In the second place, NPE is the behaviour of the current consumer when null partitions are provided (I just added the explicit validation higher in the call stack with a better message). 

That being said, I would be ok with moving to IllegalArg if we prefer, as long as we are ok with the error change regarding the old consumer.

Reply:
Maybe at the API level - illegal argument is more explicit to notify user that some arguments aren't valid.  I think Objects.requireNonNull can be at the level of the background thread as the partitions cannot be null.

Reply:
Agree with the style, but we would still be changing how the current KafkaConsumer behaves facing the user, is that ok? Current KafkaConsumer throws NPE [here](https://github.com/confluentinc/kafka/blob/2ef5b6014db0baf506df9f98eff14014a5c4c6ec/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcher.java#L292) in a call to beginningOffsets with null partitions. Just want to confirm we are ok with the change at the API level.

Reply:
Ah noted! Maybe this is something we should change here.  Would you mind adding a TODO at the prototypeAsyncconsumer level?

Reply:
The current style here is rather inconsistent. Offset related functions like this beginning/end offset throw NPE on null partitions, whereas subscribe/assign do have explicit validations and throw IllegalArg. So all that to reinforce your point that this is probably something we should change here, but it would definitely change the behaviour at the API level, so worth double-checking it with the wider team

Reply:
Yeah, we should consolidate the logic as some point. Thanks!

Reply:
I'm already on this area and given the inconsistency I'm happy to take it now that I have it in hand and then just track the follow-up with the wider team to sync on the API level change?

Reply:
Tracked [here](https://confluentinc.atlassian.net/browse/KJC-547) for later discussion. Leaving as it is for now